### ISSUE TITLE
Tracking issue fo RFC2912, transition to rust-analyzer as our official LSP

### ISSUE THREAD

**Issue by matklad** [rust-lang/rust-analyzer#4224]
###### Issues blocking providing rust-analyzer as an alternative

- [x] rust-analyzer should mostly conform to LSP protocol
  * specifically replace adHoc `"rust-analyzer.applySourceChange"` with less adHoc [SnippetTextEdit](https://github.com/microsoft/language-server-protocol/issues/724#issuecomment-619076318)
- [x] rust-analyzer should be packaged with rustup
- [ ] rust-lang.rust extension should work with both rust-analyzer and rls
  * all the code from editors/code in this repo should be moved to https://github.com/rust-lang/rls-vscode

###### Issues blocking switching to rust-analyzer as the main implementation 

- [x] rust-analyzer should fully conform to LSP protocol

###### Issues under discussion

- [ ] on the fly checking
- [ ] RAM usage
- [ ] startup time

**Comment by eminence** [comment]
Could someone briefly describe how RLS does on-the-fly checking?  I know that rust-analyzer on-the-fly checking will be implemented in its own way, but knowing how RLS does it I think could be some useful background knowledge for this thread.

**Comment by bjorn3** [comment]
RLS acts as a custom rustc driver and uses an internal compiler api ([`FileLoader`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/source_map/trait.FileLoader.html)) to provide sources directly to the parser without reading from the disk. The main problem with this, it that it is tied to a specific rustc version.

**Comment by puremourning** [comment]
> specifically, "rust-analyzer.applySourceChange" command should be opt-in

I don’t think opt-in is the way to go. Just implement these executeCommand for it like the protocol says. You can still extend the response for your vscode-only cursor position thing.

**Comment by bstaletic** [comment]
To expand on what @puremourning said, making it opt-in would leave out a lot of code actions for clients who don't explicitly opt-in. I strongly believe rust-analyzer can do better.

Instead of making `rust-analyzer.applySourceChange` opt-in, let's just (or "just") make rust-analyzer handle `executeCommand` properly. If a server sends a `CodeAction` object with a `'command'` field, that server should be ready to receive the same command through `executeCommand` request. Then the server is supposed to make a series of `applyEdit` *requests*, to let the client consume the chosen `Command`. After all of the `applyEdit` requests were, well, applied and the client responded, the server needs to send a response to `executeCommand`, which can simply be `null`.

The reason I think this shouldn't be hard, is because `rust-analyzer.applySourceChange` is already a single `WorkspaceEdit` (with an extra field that can be ignored by the clients) which is exactly what should be in an `applyEdit` request.

In other words:

1. Client makes a `textDocument/codeAction` request.
2. rust-analyzer responds with a `CodeAction`, containing a `rust-analyzer.applySourceChange` command.
3. Clients would have a choice at this point.
3.1. VSCode, for example, may continue to handle `rust-analyzer.applySourceChange` in a special way, as it needs fewer messages and thus provides better latency.
3.2. Clients that prefer protocol conformance over latency may choose to send an `executeCommand` request. Further steps only care about these clients.
4. Client sends an `executeCommand` request, containing `rust-analyzer.applySourceChange`.
5. rust-analyzer, before sending a response to `executeCommand` sends an `applyEdit` response.
5.1. This is basically just taking the `command[ 'arguments' ][ 0 ]` and putting it in `{ 'edit': <here> }`
6. The client responds with `{ 'applied': bool }`
7. rust-analyzer responds with *anything* to the `executeCommand` request.

This makes clients who want to respect the standard and clients who want to cater to rust-analyzer special cases happy.

**Comment by matklad** [comment]
Yup, sorry, that was bad wording on my part. I definitely intend to make all code-actions available to all client, with cursor positioning being a sort of "progressive enhancement". I think the plan outlined in https://github.com/microsoft/language-server-protocol/issues/724#issuecomment-619076318 should work.

**Comment by bstaletic** [comment]
I tried to implement what I was talking about in my previous comment, but I got stuck. First of all, tell me if you want me to create a separate thread for this.

I added the `ExecuteCommand` capability and am having troubles writing `handlers::handle_execute_command`. The problem is that, for step 5, I need:

1. Create a server->client request, before returning from the handler.
2. Block until the client has responded to the request.
3. Finally respond to `workspace/executeCommand`.

This is what I have written, in all its broken glory:

```rust
pub fn handle_execute_command(
    world: WorldSnapshot,
    params: req::ExecuteCommandParams,
) -> Result<Option<req::WorkspaceEdit>> {
    if params.command == "rust-analyzer.applySourceChange" {
        let arg = params.arguments[0];
        let edit : req::WorkspaceEdit = serde_json::from_value(arg).unwrap();
        if client_supports_apply_edits {
        let request_id = loop_state.next_request_id();
        let request = request_new::<req::ApplyWorkspaceEdit>(
            request_id.clone(),
            req::ApplyWorkspaceEditParams { edit: edit },
        );
        msg_sender.send(request.into())?;
        wait_for_response(); // Completely made up function name.
        return Ok(None)
        } else {
            // For clients who do not support `ApplyWorkspaceEdit` requests
            return Ok(Some(edit))
        }
    }
    return Ok(None)
}
```

Things I need:

1. `client_supports_apply_edits`: basically, client capabilities, in order to be 100% conforming.
2. `loop_state.next_request_id()`, `request_new` and `msg_sender`, in order to send the request. I'm not sure if I can/should just import `main_loop::request_new` in `handlers.rs`.
3. Some way to block until I get a response for the request I just made.

**Comment by matklad** [comment]
@bstaletic please note that we probably we'll change the precise interface here soon (I plan to overhaul this stuff next week, but it might spill longer than that), so it probably doesn't make sense to invest too much time into this right now. I'll take a closer look tomorrow, but my first thought is that I am not too excited about using server -> client applyEdit command, as it's impossible to ensure consistency with it, and it looses causality: https://github.com/microsoft/language-server-protocol/issues/663#issue-401303602

Regarding your approach, the current architecture doesn't really allow to block until a part of protocol is executed --- we kind of just have a giant explicit state machine in the main loop, and handle all events manually. I'd probably just reply to the command right away, without waiting for the client to acknowledge `applyWorkspaceEdit`  -- the client doesn't really have a way to piece the puzzle together and realise that `applyWorkspaceEdit` is connected to the executeCommand.

**Comment by puremourning** [comment]
>  the client doesn't really have a way to piece the puzzle together and realise that applyWorkspaceEdit is connected to the executeCommand.

yet, it must. in order for clients to behave sensibly (causality as you put it), we must assume that any applyEdit received twix'd an executeCommand and its reply are associated with that command.

Indeed that's exactly what we do, blocking the user while waiting for the command reply and collecting edits. This works as far as we can tell for all servers so far.

**Comment by Xanewok** [comment]
We can somewhat tick the blocking issue box:

> rust-lang.rust extension should work with both rust-analyzer and rls

[Implementation](https://github.com/rust-lang/rls-vscode/pull/793)

Also posted it in the RFC thread: https://github.com/rust-lang/rfcs/pull/2912#issuecomment-627297657

It doesn't provide exact feature parity since it doesn't work yet with messages such as `rust-analyzer.applySourceChange` and other custom extensions that are not LSP-proposed.

**Comment by matklad** [comment]
https://github.com/rust-analyzer/rust-analyzer/pull/4494 implements code action snippet capability.

**Comment by matklad** [comment]
I've transitioned all code actions to use progressive enhancement, so code actions in next release should just work with any any editor which implements `codeActionLiteralSupport`. We do not implement command-based actions, as those require custom client-side plumbing or round-about `workspace/applyEdit`. This is conforming to the protocol (we just don't emit code actions without `codeActionLiteralSupport`) and i think should be fine in practice: if client supports `workspace/applyEdit`, it shouldnt't be problematic to support suppling edit directly with code-actions.  I'll mark some comments in this thread as resolved. 

I've also started documenting all our protocol extensions [in this document](https://github.com/rust-analyzer/rust-analyzer/blob/59732df8d40dfadc6dcf5951265416576399712a/docs/dev/lsp-extensions.md). 

At this point, although we are not fully confirming to the protocol yet, I believe we've solved all major practical conformance issues.

**Comment by nikomatsakis** [comment]
Huzzah! This RFC has been merged. You can follow along at the tracking issue that @matklad created [rust-analyzer/rust-analyzer#4224](https://github.com/rust-analyzer/rust-analyzer/issues/4224).

**Comment by matklad** [comment]
Next step, shipping with rustup, hit an unexpected snag:

https://github.com/rust-lang/rust/pull/72978#issuecomment-639206392

We require too recent glibc (2.9) for file watching. Specifically, this function is not available on Rust CI builders: https://github.com/hannobraun/inotify-sys/blob/155435810601d422ccec540d5eaf90728d60ff15/src/lib.rs#L618. 

I'll look into this by making file watching optional or some such, but that's going to be a big change, so it's unlikely to land very soon. I am curious if anyone can come up with some ~~horrible~~ cute linker trick to work-around this. Can we somehow poly-fill this symbol or avoid using it altogether? 

cc @hannobraun just in case :-)

**Comment by hannobraun** [comment]
Please correct me if I'm wrong, but it looks like [`ra_vfs`](https://crates.io/crates/ra_vfs) depends on [`notify`](https://crates.io/crates/notify), which depends on [`inotify`](https://crates.io/crates/inotify), which depends on `inotify-sys`. This would mean, the problem is this call: https://github.com/hannobraun/inotify/blob/23f7ec656b8dbf573783b70d38c9988537aebf3f/src/inotify.rs#L101

If `inotify_init1` is indeed the only problem here, then this should be straight-forward to fix: Call `inotify_init` instead, then set the flags in a second call with [`fcntl`](https://linux.die.net/man/2/fcntl) ([example](https://github.com/hannobraun/inotify/blob/23f7ec656b8dbf573783b70d38c9988537aebf3f/src/inotify.rs#L288)). `notify` depends on `inotify` 0.7, so it would be necessary to also backport this change and do another 0.7.x release.

I will try to look into this soon, but if anyone wants to do the legwork here (making and testing the change), this would be highly welcome.

**Comment by matklad** [comment]
Yeah, I think the analysis is correct. Would just not calling the function would be enough though? Ie, won't we get a linker error simply because we've declared the function (I don't know how this actually works, might be totally wrong).

**Comment by hannobraun** [comment]
I'm not a linker expert by any means, but based on my experience I assume that just not calling it would fix the problem. If we're not calling it, the linker has no reason to look for it at link-time.

If I'm wrong, we can add a compile-time flag to `inotify-sys` to get rid of it completely. I'm more than happy to commit all the atrocities required to move rust-analyzer forward :-)

**Comment by mati865** [comment]
In https://github.com/rust-lang/rust/pull/72978#issuecomment-639206392 there was also `warning: lutimes is not implemented and will always fail`. That's another issue from too old GLIBC. Used by [`filetime` here](https://github.com/alexcrichton/filetime/blob/2c0e61d8b5317913cbb85912619252379ed4cc3a/src/unix/utimes.rs#L102)
Haven't checked how important it is.

**Comment by hannobraun** [comment]
I've made the change and released versions 0.8.3 and 0.7.1. Please let me know if it helps.

**Comment by matklad** [comment]
Thant worked, thanks @hannobraun !

**Comment by matklad** [comment]
Current status:

```
15:04:42|~/projects/rust-analyzer/editors/code|master✓
λ rustup component add --toolchain nightly rust-analyzer-preview
info: component 'rust-analyzer-preview' for target 'x86_64-unknown-linux-gnu' is up to date

15:04:46|~/projects/rust-analyzer/editors/code|master✓
λ rustup run nightly rust-analyzer --version
rust-analyzer f5a4a4b
```

**Comment by lzutao** [comment]
I can't add rust-analyzer now. Should I wait until tomorrow?

**Comment by matklad** [comment]
@lzutao  `rustup update nightly` perhaps?

**Comment by lzutao** [comment]
Also one could create a symlink for convenient reasons:
```bash
ln -s $(rustup which rust-analyzer ) ~/.local/bin/rust-analyzer
```

**Comment by lnicola** [comment]
Should `rustup` have a `rust-analyzer` shim?

**Comment by matklad** [comment]
@lnicola I think we need to add something else for that. 

I've opened https://github.com/rust-lang/rls-vscode/issues/812 for the last bit we need to fix before we can start the "feedback govering" phase properly.

**Comment by llucax** [comment]
Hi, on major blocker for me to adopting rust-analyzer (as a CoC plugin in (neo)vim) is the inability to update errors on the fly. This is already listed here but I didn't see any links to specific issues. Searching I found #6591, which looks like it is requesting the same. If this is correct, maybe it would be useful to link it in the TODO item in the issue description, so other people that want to monitor particular items on the TODO list can subscribe to the specific issues.

**Comment by cuviper** [comment]
> ###### Issues blocking switching to rust-analyzer as the main implementation
> 
> * [ ] rust-analyzer should fully conform to LSP protocol

Does RLS already meet this requirement?
Otherwise, I fear this goal is a case of perfect being the enemy of good.

**Comment by puremourning** [comment]
> case of perfect being the enemy of good.

what’s the point of a standard or specification if it’s not followed?

**Comment by matklad** [comment]
Yeah, that’s long been done, just forgot to tick the checkbox

**Comment by matklad** [comment]
Note that the real blocker today is non-technical. We need to move rust-analyzer over to rust-lang project before it can be made official. We are slowly but steadily doing that with the rust-lang core team.

**Comment by cuviper** [comment]
@puremourning 

> what’s the point of a standard or specification if it’s not followed?

My point was to consider this relative to RLS that it is replacing.

**Comment by Xanewok** [comment]
Since https://blog.rust-lang.org/2022/02/21/rust-analyzer-joins-rust-org.html has been announced and it seems that Rust Analyzer has joined the Rust org (does it mean that the repository will move?), I'd gladly move or replace the code for the current VSCode extension with the Rust Analyzer VSCode extension (maybe with an added option to run RLS if already installed by rustup) to quickly move things forward.

To give some more context from my side: I've obviously dropped the ball with the migration tracked at https://github.com/rust-lang/vscode-rust/issues/812. I was and am still struggling with mental health issues and it became apparent that I cannot keep up with my existing open source obligations (at least I viewed them as such) and couldn't maintain the outlined pace of work nor even communicate that properly. I'm really sorry for that.

I'm very happy that the Rust IDE story is getting better and better with every week and I'll happily coordinate with adding new maintainers to the official Rust VSCode extension and with any work required to make the switch (should I CC any specific RA team?). Thanks for all your hard work on the Rust Analyzer and other IDE-related bits!

**Comment by digitarald** [comment]
I work on VS Code and a few team members also noted how much better Rust editing is with _rust-analyzer_. I dug deeper as VS Code had a built-in recommendation for the _rust-lang_ extension and found this issue.

As one fix the upcoming VS Code 1.66 will start recommending rust-analyzer. Until the extensions are unified this seems the best experience for users. Feel free to contact me if you need help with the transitioning in VS Code or Marketplace.

**Comment by Veykril** [comment]
As `rust-analyzer` has transitioned to the rust-lang org we can close this issue 🎉

There was a slight change of plans regarding the VSCode extensions, as instead of merging the two VSCode extensions as described in https://github.com/rust-lang/vscode-rust/issues/812, https://github.com/rust-lang/vscode-rust is now being deprecated in favor of the `rust-analyzer` vscode extension which also has now moved publisher to `rust-lang` on the [marketplace](https://marketplace.visualstudio.com/items?itemName=rust-lang.rust-analyzer). Likewise `RLS` is being sunset https://github.com/rust-lang/rust/pull/100863 now, although that was outlined in the RFC anyways.
